### ISSUE TITLE
[$150] Notifications about posts in Phases are not handled correctly

### ISSUE THREAD

**Issue by maxceem** [appirio-tech/connect-app#2575]
When we make a comment in some Phase `Discussions` tab other users are getting a notification about it. 

### Expected behavior
- When we click such notifications it has to 
   - go to `Project Plan` tab
   - scroll to appropriate Phase
   - expand Phase
   - open`Discussions` tab
   - make sure that related comment is visible
- The notification has to be marked as `read` when we open a `Discussions` tab in the particular Phase (works now) but **NOT** in other conditions (doesn't work now).

### Actual behavior
- When we click such notifications it leads to the `Dashboard` tab where we do not always have a phase with the related comment. And even if we have a neccessary phase here it is not expanded.
- This notification is marked as `read` when we come to `Dashboard` tab.

**Comment by maxceem** [comment]
This issue is basically a duplicate of https://github.com/appirio-tech/connect-app/issues/2175.
So we depend on the backend changes here.

**Comment by vikasrohit** [comment]
@maxceem where are we blocked for this one now? Can we unblock it and include it in our CF challenge?

**Comment by maxceem** [comment]
Unblocked and added to cf-pipeline.

**Comment by maxceem** [comment]
At the moment when we click notification about post for the phase we:
- navigated to the phase tab
- scroll to the phase
- open Discussion tab

The only thing we don't do at the moment is:
- scroll to a particular comment on the phase as it's done for comments on the dashboard
 
To accomplish it we should do several things:
- change `PHASE_POST` (https://github.com/appirio-tech/connect-app/blob/dev/src/routes/notifications/constants/notifications.js#L20) URL to `/projects/{{projectId}}/plan/#phase-{{phaseId}}-posts-{{postId}}` so phase post notifications lead to the particular phase post instead of just to the tab
- define id of phase posts to math the new URL here https://github.com/appirio-tech/connect-app/blob/cf16/src/projects/detail/containers/PhaseFeedHOC.jsx#L117
- we should update code to expand `Discussions` tab if necessary https://github.com/appirio-tech/connect-app/blob/cf16/src/projects/detail/containers/ProjectPlanContainer.jsx#L64 so when it sees a link in URL which looks like `/projects/{{projectId}}/plan/#phase-{{phaseId}}-posts-{{postId}}` it would open `Discussion` tab automatically. Now it already works for links like `/projects/{{projectId}}/plan/#phase-{{phaseId}}-posts`. NOTE: it should still work for `/projects/{{projectId}}/plan/#phase-{{phaseId}}-posts` URLs as we use such URLs to link to the phase Discussion tab in the sidebar.
- at the moment when we click links like `#phase-{{phaseId}}-posts` there is automatic smooth scrolling to the phase, see demo https://monosnap.com/file/gocBgWt1gdUxcLV318UygVNaUoZqWz
   - this automatic smooth scrolling is handled by `react-scroll` library: it tries to find element an element with `id` which is mentioned in browser URL hash and if there is such, it scrolls to it smoothly.
  - after we change URL for phase posts to `/projects/{{projectId}}/plan/#phase-{{phaseId}}-posts-{{postId}}` such posts would be not rendered at the moment when we click such URL. I think `react-scroll` library can handle some cases when elements are not yet rendered and it will scroll as soon as they are rendered. But in some cases it may fail to do so. If for our case it will fail do so we can still trigger smooth scrolling using https://github.com/appirio-tech/connect-app/blob/cf16/src/components/ScrollToAnchors.jsx#L40 HOC or `scrollToHash` method.

**Comment by applehit16** [comment]
I'll take this one

**Comment by maxceem** [comment]
@applehit16 it's on you since you commented as per our Bug Bash rules. To keep it clear I can assign you to this issue in the UI. For this, please accept the invitation to this repository which you should get to your email.

**Comment by applehit16** [comment]
Since the dev api was down yesterday, and I didn't find myself have time to look into this. Feel free to unassign from me.

**Comment by sumitdaga** [comment]
Can you please assign this to.me @maxceem ?

**Comment by maxceem** [comment]
Thank you for update @applehit16.

@sumitdaga it's on you.

**Comment by sumitdaga** [comment]
@maxceem  
by post we mean scroll to a specific comment right ? 
but a specific comment in the discussions tab has id like `commnet-{id}`  ...isn't it ?

**Comment by maxceem** [comment]
> by post we mean scroll to a specific comment right ?

right. In this task we are talking about posts/comments on the `Project Plan` tab, which are placed on the `Discussion` tabs of the Phases.

![image](https://user-images.githubusercontent.com/146016/56029768-b90cce80-5d4d-11e9-8e64-fe856bb4e7a7.png)

> but a specific comment in the discussions tab has id like commnet-{id} ...isn't it ?

on the discussion tab - correct. But on the Project Plan tab comments should have ID like `phase-{{phaseId}}-posts-{{postId}}`

**Comment by sumitdaga** [comment]
> on the discussion tab - correct  
you mean dashboard tab

see this though
![Workspace 1_073](https://user-images.githubusercontent.com/7769148/56031471-34b24f80-5d3d-11e9-958b-48496679c84f.png)

**Comment by maxceem** [comment]
Right, we use the same component to render comments, so currently, it uses the same ID in both places. So there are two ways to go:
- in the way described above we should changes IDs of comments on the Project Plan tab to be like `phase-{{phaseId}}-posts-{{postId}}` while keeping them as they are on the Dashboard tab. And after use these links to navigate to them.
- alternatively, we can keep IDs of comments as they are and use links like `/projects/{{projectId}}/plan/#comments-{postId}` to navigate to the comments. But this way has a  drawback: we should wait until comments are loaded to detect which phase they belong to as such link doesn't have information about the particular phase.  

So it feels like it's easier to change comment IDs on the project plan tab.
Let me know what you think.

**Comment by sumitdaga** [comment]
I agree chaning comment IDs seems more fit ...
i have tried it and seems to work  
so i am passing a prop `commentAnchorPrefix` from the project plans tab ...and if it is defined in the [Comment Component](https://github.com/appirio-tech/connect-app/blob/dev/src/components/ActionCard/Comment.jsx#L62) it will use it or else fall back to `comment-{id}`  
sounds ok? i will push in some time so you can also check

**Comment by sumitdaga** [comment]
btw i missed this  
> define id of phase posts to math the new URL here https://github.com/appirio-tech/connect-app/blob/cf16/src/projects/detail/containers/PhaseFeedHOC.jsx#L117  

this permalink does not seem to be used anywhere ?  
should this be used instead of a new prop like i suggested above ?

**Comment by maxceem** [comment]
> sounds ok?

yeah, sound good.

> this permalink does not seem to be used anywhere ?

could be. just make sure that when we click links of the phases in the left sidebar they still work good and navigate to discussions tabs of phases, https://monosnap.com/file/T0GtZgnAO9q1vUHiAa4D55J1roD2em

**Comment by sumitdaga** [comment]
@maxceem  PR at #2948  
let me know if its ok?

**Comment by sumitdaga** [comment]
and btw i think the url should be  
`/projects/{{projectId}}/plan#phase-{{phaseId}}-posts-{{postId}}` and not  
`/projects/{{projectId}}/plan/#phase-{{phaseId}}-posts-{{postId}}`   
i have updated the PR

**Comment by maxceem** [comment]
Everyone who worked on the issues during Bug Bash 16.

Please, don't forget to submit a file with a list of issues you were working on to the challenge page https://www.topcoder.com/challenges/30088391

This would speed the payments.

Thank you.

cc @mfikria @rashmi73 @sumitdaga @PrakashDurlabhji @Gauravseta

**Comment by RishiRajSahu** [comment]
validated working fine in dev !
